---
title: "Aldrich_final_knit.rmd"
output: github_document 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)

#package installation 
library(ggplot2)
library(tidyverse)
library(GenomicRanges)
library(dplyr)
library(pheatmap)

#dependencies for ComplexHeatmap 
#install.packages("https://cran.r-project.org/src/contrib/Archive/rjson/rjson_0.2.15.tar.gz", repos=NULL, type="source")
#install.packages("https://cran.r-project.org/src/contrib/Archive/GetoptLong/GetoptLong_1.0.3.tar.gz", repos=NULL, type="source")
#BiocManager::install("ComplexHeatmap")

library(ComplexHeatmap)


#source util functions 
source("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/util/plotting_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/util/_setup.R")
source("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/util/my_class_functions.R")

```

#Objective 
The ENCODE database is the largest ChIP-seq database available and provides the opportunity for analysis of DNA binding proteins (DBPs) in the same cell state. Here, we leverage ENCODE data to ask questions about and make comparisons between 430 DNA DBPs. Some guiding questions for this project include: 


(i) What are the number of peaks and genome coverage for each DBP? 

(ii) What are the binding preferences for promoters, gene-bodies and intergenic genomic regions? 

(iii) What are the similarities and differences across DBPs based on their genome-wide binding profiles genome-wide? 


#Methods
Analysis will take place over five steps:
1. Create Consensus Peaks 
2. Plot Consensus Peaks
3. Cluster Peaks
4. RNAseq Analysis 
  4.1: Compare Binding and Expression
  4.2: Superbinder Expression


#Step 1: Load Consensus Peaks Annotation File and Create Consensus Peaks
```{r load-consensus-peaks}

load("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/results/peak_features.RData", verbose = T)


load("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/results/filtered_consensus_list.RData")


```

#Step 2: Plot Consensus Peaks 

## Peaks per dbp
```{r plotting peak features}

# First let's look at a histogram of peak#/DBP
 ggplot(num_peaks_df, aes(x = num_peaks)) + 
  geom_histogram(bins = 70)
# saving
ggsave("figures/num_peaks_hist.pdf")

```
###Result: After filtering for proteins that have at least 1000 peaks, we will move forward with 430 DBPs


## Plot num_peaks versus genome coverage.
```{r peaks vs coverage}

ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length)) +
  geom_point() + 
  geom_smooth(method = "gam", se = TRUE, color = "black", lty = 2)+
  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")

# saving 
ggsave("figures/peak_num_vs_coverage.pdf")


```
###Result: There is a linear relationship between number of peaks and total coverage


## Plot number of peaks on promoters

```{r number of DBPS on promoters}

ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_promoters)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping promoters") +
  ggtitle("Relationship Between Number of DBP Peaks and Promoter Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE, formula = 'y ~ x',
              color = "#a8404c") +
  ylim(0,60100) +
  xlim(0,60100)

ggsave("figures/peak_num_vs_promoter_coverage.pdf")

# peaks you stop increasing binding to promoters.
# maybe it leaks into the gene body let's check
```
###Result: As the number of peaks per DBP increases, the number of peaks overlapping promoters starts to level off around 20,000. This behavior suggests binding saturation.


## Quantify peak coverage on gene bodies
```{r peak coverage on gene bodies}

ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_genebody)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping genes") +
  ggtitle("Relationship Between Number of DBP Peaks and Gene Body Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=F, formula = 'y ~ x',
              color = "#a8404c") +
  ylim(0,60100) +
  xlim(0,60100)

# saving
ggsave("figures/4_peak_num_vs_gene_body_coverage.pdf")
```
###Result: The linear relationship between peaks per DBP and number of peaks overlapping with genes supports that gene bodies explain almost all the places of binding in the genome. 


## Plot density of binding events (DBPs bound per promoter)
```{r density plot of DBP localization events}

ggplot(peak_occurence_df, aes(x = number_of_dbp)) +
geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") 

# saving
ggsave("figures/num_binding_events_per_promoter.pdf")

```
###Result: There are two types of promoter binding, evident by the two peaks in density, "normal" and "superbinders". Normal promoters have ~100DBPs and the superbinders have ~200DBPs. Subsequent analysis is needs to analyze if these superbinders interact with mRNA or lncRNA 


## First compare lncRNA versus mRNA promoter binding for all promoter types (normal and superbinders)
```{r lncrna vs mrna promoter binding}
num_peaks_dfl <- num_peaks_df %>%
  dplyr::select(-peaks_overlapping_promoters) %>%
  pivot_longer(cols = peaks_overlapping_lncrna_promoters:peaks_overlapping_mrna_promoters,
               names_to = "gene_type",
               values_to = "peaks_overlapping_promoters") %>%
  mutate(gene_type = gsub("peaks_overlapping_", "", gene_type))

# plotting
ggplot(num_peaks_dfl, aes(x = num_peaks, y = peaks_overlapping_promoters, 
                         col = gene_type)) +
         geom_point() +
         geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se = FALSE, formula = "y ~ x") + 
  scale_color_manual(values = c("#a8404c", "#424242"))+
  xlab("Peaks per DBP") +
  ylab("Peaks Overlapping Promoters") +
  ggtitle("Number of DBP Peaks and Promoter Overlaps")

# saving
ggsave("figures/peaks_overlaps_relationship_by_gene_type.pdf", height = 5, width = 8)

```
###Results: 
###TODO: interpret ###


## Now filter peak_occurence_df for superbinders and compare again mRNA vs lncRNA
```{r}

superbinders_df <- filter(peak_occurence_df, number_of_dbp > 185)

lnc_superbinders <- filter(superbinders_df, gene_type == "lncRNA")
percent_lnc_superbinders <- (nrow(lnc_superbinders) / nrow(superbinders_df))*100

mRNA_superbinders <- filter(superbinders_df, gene_type == "protein_coding")
percent_mRNA_superbinders <- (nrow(mRNA_superbinders) / nrow(superbinders_df))*100

```
###Results: ~22% of superbinders bind to lncRNA, while ~79% of superbinders bind to mRNA.


## Plot promoters without binding events
```{r prmoters with out binding events}

unbound_promoters <- peak_occurence_df %>% 
  filter(peak_occurence_df$number_of_dbp < 1)

percent_unbound <- (nrow(unbound_promoters) / nrow(peak_occurence_df))*100

write_csv(unbound_promoters, "results/unbound_promoters.csv")

```
###Results: 9,448 promoters do not have binding events. This represents ~25% of all promoters.


#Step 3: Cluster Peaks

##Load Promoter Peak Occurence df
```{r load promoter peak occurence}
#load("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/results/peak_occurence_dataframe.RData")

#not necessary bc promoter_peak_occurence_df is already in peak_features.RData, but another source for peak_occurence_df 

```

##Create distance matrix & dendrogram
```{r distance matrix and dendrogram}

# creating distance matrix
peak_occurence_dist <- dist(promoter_peak_occurrence_matrix, method = "binary")

# clustering distance values
bin_hier <- hclust(peak_occurence_dist, method = "complete")

# Dendrogram of binding profiles by promoter (not binding profile - below)
ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3,
                       theme_dendro = TRUE) +
   coord_flip() +
   scale_y_continuous() +
   scale_x_continuous(position = "top") +
   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) +
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )

ggsave("figures/promoter_overlap_dendrogram.pdf")

```

# Using profile_tss for all 430 DBPs
# ! this takes ~45 min !
```{r metaplot DF of binding profiles over promoter}

# establishing DF
metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate DF 
for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  metaplot_df <- bind_rows(metaplot_df, tmp_df)
  
}

# saving
saveRDS(metaplot_df, "results/metaplot_df_final.rds")

```
  

## Create distance matrix of binding profile correlations
```{r scaling and plotting dendrogram of binding similarity by promoter}
metaplot_filtered_matrix <- metaplot_df %>% 
  pivot_wider(names_from = x, values_from = dens) %>%
  column_to_rownames("dbp") %>%
  as.matrix()
mm_scaled <- t(scale(t(metaplot_filtered_matrix)))
metaplot_hclust <- hclust(dist(mm_scaled), method = "complete")

# plotting relationship between binding profiles
plot(metaplot_hclust)
pdf("figures/tss_profile_dendrogram.pdf", height = 10, width = 27)
par(cex=0.3)
plot(metaplot_hclust)
dev.off()
```



## Make heat map of binding profile over +/- 1kb TSS window
```{r heatmap of profile over TSS (+/- 1Kb)}
# setting up PDF function for saving
pdf("figures/tss_profile_heatmap.pdf", height = 35, width = 10)

# complex heatmap of mm_scaled
Heatmap(mm_scaled, cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()

```

## Cluster based on lncRNA  
```{r lncRNA promoter clustering}

# load lncRNA promoters annotations 
#lncrna_mrna_promoters <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/05_R_analyses/01_peak_features/results/gene_annotations/lncrna_mrna_promoters.gtf")

# split into lncRNA and mRNA
lncrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "lncRNA"]

mrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "protein_coding"]

# separate peak_occurrence_matrix in lncRNA and mRNA
lncrna_peak_occurence <- promoter_peak_occurrence_matrix[,lncrna_promoters$gene_id]

# hclust:
bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))

# plot:
ggdendro::ggdendrogram(bin_hier_lncrna, rotate = T,  size = 3)
 
# save 
ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 49, width = 6)

```
## Cluster based on mRNA 
```{R mRNA promoter clustering}
mrna_peak_occurence <- promoter_peak_occurrence_matrix[,mrna_promoters$gene_id]

# get the distance matrix for only mRNA promoters  
bin_hier_mrna <- hclust(dist(mrna_peak_occurence, method = "binary"))
 
# plot with ggdendro
ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)

# save
ggsave("figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)

```

## Investigate patterns of ChIP binding profiles relative to TSS: separate into several clusters
```{r how many groups are there?}

# Need to know how many DBPs have different patterns.
# Using guess and check, after 6 we get a new group of 10 (after that not much changes)
clusters <- cutree(metaplot_hclust, k=6)
table(clusters)

# plot cluster 1
dev.new()
pdf("figures/cluster_1_heatmap.pdf", height = 35, width = 10)

Heatmap(mm_scaled[clusters == 1,], cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()


# cluster 2
dev.new()
pdf("figures/cluster_2_heatmap.pdf", height = 35, width = 10)

Heatmap(mm_scaled[clusters == 2,], cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()


# cluster 3
dev.new()
pdf("figures/cluster_3_heatmap.pdf", height = 35, width = 10)

Heatmap(mm_scaled[clusters == 3,], cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()


# cluster 4
dev.new()
pdf("figures/cluster_4_heatmap.pdf", height = 35, width = 10)

Heatmap(mm_scaled[clusters == 4,], cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()

# looks like TSS depletion and histone mods

# cluster 5 only 1 DBP
names(clusters[5])
#AKAP8

# cluster 6
dev.new()
pdf("figures/cluster_6_heatmap.pdf", height = 35, width = 10)
Heatmap(mm_scaled[clusters == 6,], cluster_columns = FALSE, border = TRUE, use_raster = TRUE, column_gap = unit(0, "mm"), row_names_gp = gpar(fontsize = 7))
graphics.off()


```
###Results: In cluster 1, TSS is enriched a bit broader than cluster 2. Cluster 3 has very narrow binding over TSS relative to clusters 1 and 2. Cluster 4 has evidence of TSS depletion, potentially due to histones/histone modifications. Cluster 5 is less interesting because it only contains 1 DBP, AKAP8. Cluster 6 is also small with only 3 DBPs, and like cluster 4 shows TSS depletion/histone modifications. 



# establishing lncRNA and mRNA promoters (+/- 1kb)
```{r create lncRNA and mRNA promoters }

# creating promoters just in case:
lncrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "lncRNA"]
mrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "protein_coding"]

```

## Make metaplots for each DBP by lncRNA and mRNA promoters
```{r}

#setting up lncrna DF.
lncrna_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())


# for loop to populate DF with overlap density in lncrna promoters
for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = lncrna_promoters)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  lncrna_metaplot_df <- bind_rows(lncrna_metaplot_df, tmp_df)
  
}

# saving
saveRDS(lncrna_metaplot_df, "results/lncRNA_metaplot_df_final.rds")

# now for mRNAs 
mrna_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate mRNA_metaplot
for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = mrna_promoters)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  mrna_metaplot_df <- bind_rows(mrna_metaplot_df, tmp_df)
  
}

# saving mRNA metaplots
saveRDS(mrna_metaplot_df, "results/mrna_metaplot_df_final.rds")


# now adding the information of gene type
mrna_metaplot_df$gene_type <- "mRNA"
lncrna_metaplot_df$gene_type <- "lncRNA"
combined_metaplot_profile <- bind_rows(mrna_metaplot_df, lncrna_metaplot_df)

# saving
saveRDS(mrna_metaplot_df, "results/metaplot_df_final.rds")



# pdf(file = "figures/mega_plot_test.pdf")
# ggplot(combined_metaplot_profile, 
#        aes(x = x, y = dens, color = gene_type )) +
#   geom_vline(xintercept = 0, lty = 2) + 
#   geom_line(size = 1.5) + 
#   facet_wrap(dbp ~ ., scales = "free_y") +
#   ggtitle("Promoter Metaplot") + 
#   scale_x_continuous(breaks = c(-1000, 0, 1000),
#                      labels = c("-1kb", "TSS", "+1kb"),
#                      name = "") + 
#   ylab("Peak frequency") +
#  scale_color_manual(values = c("#424242","#a8404c"))

# saving
# ggsave("figures/mega_meta_plot_lncRNA-mRNA.pdf", width = 49, height = 12)


#since that plot renders in a way that's hard to see, here's a screenshot of the pdf:

#first half of mega_meta_plot_lncRNA-mRNA.pdf: 
knitr::include_graphics("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/figures/mega_meta_1.png")

#second half of mega_meta_plot_lncRNA-mRNA.pdf:
knitr::include_graphics("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/figures/mega_meta_2.png")


```
###Results: There are a couple really interesting proteins that bind around the promoter, but whose binding on the promoter is depleted: EZH2 and H2AFZ. EZH2 encodes histone methyltransferase, while H2AFZ encodes a member of the histone 2A family. 

###As expected, histones have broad U-shaped plots where the least binding occurs around the promoter. 


## Metaplot analysis of DBPs on superbinders vs regular promoters... a noble effort but a failed attempt
```{r metaplot analysis of DBPs on superbinders vs regular promoters}
# # filtering super binders
# peak_occurence_df <- peak_occurence_df %>%
#   mutate(superbinder = peak_occurence_df$number_of_dbp > 200)
# 
# # setting column of superbinders
# peak_occurence_df <- peak_occurence_df %>%
#   mutate(superbinder2 = ifelse(peak_occurence_df$superbinder ==T, "super_binder", "not_super_binder"))
# 
# # subset super binders
# super_binder_promoters <- subset(peak_occurence_df, superbinder2 == "super_binder")
# super_binder_promoters <- dplyr::select(super_binder_promoters, "gene_id")
# 
# # subset non super binders
# non_super_binder_promoters <- subset(peak_occurence_df, superbinder2 == "not_super_binder")
# super_binder_promoters <- dplyr::select(non_super_binder_promoters, "gene_id")
# 
# # subset lnc & mRNA promoters by superbinders (to get into a GRanges object)
# super_binder_gr <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_id %in% super_binder_promoters$gene_id]
# non_super_binder_gr <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_id %in% non_super_binder_promoters$gene_id]
# 
# # making empty df to populate for super binders
# highbinder_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
# 
# # for loop to populate high binder metaplot
# suppressWarnings({
# for(i in 1:length(filtered_consensus_list)) {
#  # print(names(filtered_consensus_list)[[i]])
#   tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = super_binder_gr)
#   tmp_df$dbp <- names(filtered_consensus_list)[[i]]
#   highbinder_metaplot_df <- bind_rows(highbinder_metaplot_df, tmp_df)
#  
# }
# })
# 
# # saving high binder metaplots
# write_rds(highbinder_metaplot_df, "results/highbinder_metaplot_df_final.rds")
# 
# # make metaplot for high binder profiles
# suppressWarnings({
# ggplot(highbinder_metaplot_df,
#        aes(x = x, y = dens)) +
#   geom_vline(xintercept = 0, lty = 2) +
#   geom_line(size = 1.5) +
#   facet_wrap(dbp ~ ., scales = "free_y") +
#   ggtitle("High Binder Metaplot") +
#   scale_x_continuous(breaks = c(-1000, 0, 1000),
#                      labels = c("-1kb", "TSS", "+1kb"),
#                      name = "") +
#   ylab("Peak frequency") +
#  scale_color_manual(values = c("#424242","#a8404c"))
# })
#  
# # saving high binder metaplot
# ggsave("figures/highbinder_metaplot.pdf", width = 49, height = 12)
# 
# # making empty df to populate for non super binders
# lowbinder_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
# 
# # for loop to populate non super binder metaplot
# suppressWarnings({
# for(i in 1:length(filtered_consensus_list)) {
#  # print(names(filtered_consensus_list)[[i]])
#   tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = non_super_binder_gr)
#   tmp_df$dbp <- names(filtered_consensus_list)[[i]]
#   lowbinder_metaplot_df <- bind_rows(lowbinder_metaplot_df, tmp_df)
#  
# }
# })
# 
# # saving non super binder binder metaplots
# write_rds(lowbinder_metaplot_df, "results/lowbinder_metaplot_df_final.rds")
# 
# # make metaplot for non super binder binder profiles
# suppressWarnings({
# ggplot(lowbinder_metaplot_df,
#        aes(x = x, y = dens)) +
#   geom_vline(xintercept = 0, lty = 2) +
#   geom_line(size = 1.5) +
#   facet_wrap(dbp ~ ., scales = "free_y") +
#   ggtitle("Low Binder Metaplot") +
#   scale_x_continuous(breaks = c(-1000, 0, 1000),
#                      labels = c("-1kb", "TSS", "+1kb"),
#                      name = "") +
#   ylab("Peak frequency") +
#  scale_color_manual(values = c("#424242","#a8404c"))
# })
#  
# # saving low binder metaplot
# ggsave("figures/lowbinder_metaplot.pdf", width = 49, height = 12)
# 
# # adding gene_type
# lowbinder_metaplot_df$gene_type <- "non_super_binder"
# highbinder_metaplot_df$gene_type <- "superbinder"
# 
# # making combined metaplot
# combined_super_binder_metaplot_profile <- bind_rows(lowbinder_metaplot_df, highbinder_metaplot_df)
# 
# ggplot(combined_super_binder_metaplot_profile,
#        aes(x = x, y = dens, color = gene_type )) +
#   geom_vline(xintercept = 0, lty = 2) +
#   geom_line(size = 1.5) +
#   facet_wrap(dbp ~ ., scales = "free_y") +
#   ggtitle("Promoter Metaplot") +
#   scale_x_continuous(breaks = c(-1000, 0, 1000),
#                      labels = c("-1kb", "TSS", "+1kb"),
#                      name = "") +
#   ylab("Peak frequency") +
#  scale_color_manual(values = c("#424242","#a8404c"))
# 
# # saving combined binder metaplot
# ggsave("figures/high_and_low_binder_metaplot.pdf", width = 49, height = 12)


#Unfortunately, never resolved the unused argument error! 

```

#Step 4: RNAseq Analysis
##Step 4.1: Compare Binding and Expression 

##Read in public RNA seq data
```{read in public RNA seq data}
#ENCODE report on experimental accessions used (fractionated and total RNA HEPG2)
wget -O samples.txt “https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=Transcription&assay_slims=Transcription&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&biosample_ontology.term_name=HepG2&biosample_ontology.classification=cell+line&assay_title=total+RNA-seq&files.read_length=50&limit=all&advancedQuery=date_released:[2009-01-01%20TO%202021-12-31]”

#and the file names:
wget -O files.txt “https://www.encodeproject.org/batch_download/?type=Experiment&status=released&assay_slims=Transcription&assay_slims=Transcription&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&biosample_ontology.term_name=HepG2&biosample_ontology.classification=cell+line&assay_title=total+RNA-seq&files.read_length=50&limit=all&advancedQuery=date_released:[2009-01-01%20TO%202021-12-31]”

```

## Read in TPM values from Salmon for our analyses
```{r reading in salmon Salmon TPMs}
# read in the sample sheet
samplesheet <- read_rds("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/05_R_analyses/05_RNAseq/01_differential_expression/results/final_samplesheet.rds")

# reading in salmon tpm
salmon_tpm <- read.csv("/scratch/Shares/rinnclass/CLASS_2023/data/data/rnaseq/results/salmon/salmon_merged_gene_tpm.csv")

# make sure TPM table is in same order as samplesheet
tpm <- salmon_tpm %>% 
  pivot_longer(cols = 2:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")

```

## Abundance of genes in each cellular fraction
```{r TPM of genes in each fraction}

# convert tpm DF into a matrix
tpm_matrix <- tpm %>% 
  column_to_rownames("gene_id") %>%
  as.matrix()
tpm_scaled <- t(scale(t(tpm_matrix)))
tpm_scaled <- tpm_scaled[complete.cases(tpm_scaled),]


# plotting
new.env()
pdf("figures/heatmap_expression.pdf", height =49, width = 12)
pheatmap::pheatmap(tpm_scaled, show_rownames = FALSE)
graphics.off()

```

###Result: Most RNAs are nuclear! Next most common fraction is insoluble cytosolic. 


## Plotting binding versus expression
```{r DBP promoter binding versus total RNA expression}

#join promoter features df and tpm 
#promoter_features_with_tpm <- merge(x=promoter_features_df,y=tpm, 
#             by="gene_id", all.x=TRUE)

load("/scratch/Shares/rinnclass/CLASS_2023/emal2471/CLASS_2023/CLASSES/06_final_going_global/analysis/5_1_2023.RData")

ggplot(promoter_features_with_tpm, 
            aes(y = log2(tpm_homo_sapiens_hepg2 + 0.001), 
                x = number_of_dbp, 
                color = gene_type)) + 
  geom_point(alpha = 0.2) +
  geom_smooth(method = "lm") +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

ggsave("figures/binding_vs_expression_total_rna.pdf")
```
###Result: It is already known that more binding events at a promoter is indicative of more abundant expression. It is good validation that we see a linear relationship between the number of DBPs and expression. However, there is a subset of genes with lower expression given the number of DBPs-- following analysis will see what fraction of RNA causes this pattern 


## Binding versus nuclear expression

```{r binding versus nuclear expression}

# plot nuclear RNA abundance versus #DBPs bound to their promoter
ggplot(promoter_features_with_tpm, 
            aes(y = log2(tpm_homo_sapiens_nuclear_fraction + 0.001), 
                x = number_of_dbp, 
                color = gene_type)) + 
  geom_point(alpha = 0.2) +
  geom_smooth(method = "lm") +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Nuclear Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

  # saving figure
  ggsave("figures/nuclear_expression-vs-promoter_binding.pdf")
```
###Result: looks very similar to total RNA binding versus expression


## Binding versus cytoplasmic expression
```{r binding versus cytoplasmic expression}


ggplot(promoter_features_with_tpm, 
            aes(y = log2(tpm_homo_sapiens_cytosolic_fraction + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(alpha = 0.2) +
  geom_smooth(method = "lm") +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Cytoplasmic Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 
  # saving figure
  ggsave("figures/cytoplasmic_expression-vs-promoter_binding.pdf")
  
```
###Results: we see low abundance! Linear trend seen before is also seen here, but it's due to low-expression binding events. 


## lncRNA versus mRNA expression in total RNA
```{r determining lncRNA and mRNA expression levels in total RNA}

ggplot(promoter_features_with_tpm, aes(x = log2(tpm_homo_sapiens_hepg2 + 0.01), color = gene_type))+
  geom_density()

# saving figure
ggsave("figures/mrna_lncrna_tpm_total_rna.pdf")

#lncRNAs are typically more nuclear, repeat with nuclear fraction: 
ggplot(promoter_features_with_tpm, aes(x = log2(tpm_homo_sapiens_nuclear_fraction + 0.01), color = gene_type))+
  geom_density()

# saving figure
ggsave("figures/mrna_lncrna_tpm_nuclear.pdf")
```
###Result:Good validation that lncRNAs exhibit lower expression levels compared to mRNAs. However, we see unusually high levels of mRNA with nuclear expression, which is unexpected... potentially, hepG2 cells have more DNA-protein interactions sites than a linear relation with its expression levels would suggest


#Step 4.2: Superbinder Expression
##Determine percentage of superbinders that have tpm of less than .1 and compare this to normal binders

```{r filter superbinders with tpm of less than .1 percent}
# filter promoter features 
promoter_features_with_tpm <- promoter_features_with_tpm %>%
  mutate(superbinder = promoter_features_with_tpm$number_of_dbp > 200)

# make a new column in promoter_features_df for superbinders
promoter_features_with_tpm <- promoter_features_with_tpm %>% 
  mutate(superbinder2 = ifelse(promoter_features_with_tpm$superbinder ==T, "superbinder", "not_superbinder"))

# compare superbinder expression...start with total RNA
ggplot(promoter_features_with_tpm, aes(x = log2(tpm_homo_sapiens_hepg2 + 0.01), color = superbinder2))+
  geom_density()

ggsave("figures/superbinder_tpm_total_rna.pdf")


#now for superbinders that have less than 0.1% tpm
superbinder_promoter_features_with_tpm <- promoter_features_with_tpm[promoter_features_with_tpm$superbinder2 == "superbinder", ]

#make a new column for hepg2 tpm < 0.001 (0.1%)
superbinder_promoter_features_with_tpm <- superbinder_promoter_features_with_tpm %>%
  mutate(less_than_0.1_tpm = superbinder_promoter_features_with_tpm$tpm_homo_sapiens_hepg2 < 0.001)

#make it easier to plot
superbinder_promoter_features_with_tpm <- superbinder_promoter_features_with_tpm %>%
  mutate(superbinder_expression = ifelse(superbinder_promoter_features_with_tpm$less_than_0.1_tpm ==T, "lower than 0.1% expression superbinder", "greater than 0.1% expression superbinder"))

ggplot(superbinder_promoter_features_with_tpm, aes(x = log2(tpm_homo_sapiens_hepg2 + 0.01), color = superbinder_expression))+
  geom_density()

ggsave("figures/superbinder_tpm_less_than_0.1_percent.pdf")

low_exp_superbinders <- promoter_features_with_tpm[superbinder_promoter_features_with_tpm$superbinder_expression == "lower than 0.1% expression superbinder", ]

nrow(low_exp_superbinders)
#507

nrow(superbinder_promoter_features_with_tpm)
#11689

percentage_less_than_0.1_percent <-  (nrow(low_exp_superbinders) / nrow(superbinder_promoter_features_with_tpm))*100
#4.337

```
###Results: There are 507 superbinders with a tpm of less than 0.1%. Of the 11,689 superbinders total, this represents ~4.34 percent. 







